### PR TITLE
feat(options_validator): add overrideOptions functionality

### PR DESCRIPTION
**Pull request by kvwalker**
Fixes [NODE-1673](https://jira.mongodb.org/browse/NODE-1673)

### CODE REVIEW

File: lib/options_validator.js [hunk 1]
```diff
@@ -91,6 +92,23 @@ function validationFunction(optionsSchema, providedOptions, validationOptions) {
     }
   });
 
+  if (overrideOptions) {
+    Object.keys(overrideOptions).forEach(optionName => {
```
[the remark this answers is not in the document]
Comment (kvwalker) [reply lib/options_validator.js]: Aren't there some cases where we add options even if they aren't provided? In `dropDatabase`, we ensure a primary readPreference:
```
const finalOptions = Object.assign({}, this.s.options, {
  readPreference: ReadPreference.PRIMARY
});
```

[the remark this answers is not in the document]
Comment (mbroadst) [reply lib/options_validator.js]: hmm good point, I was going to say we should move that to a `default`, but its not a configurable option. Okay, in that case where you bail out above in this function if the validation level is `NONE`, then you still need to apply the overrides.

File: lib/options_validator.js [hunk 2]
```diff
@@ -27,14 +27,20 @@ function createValidationFunction(optionsSchema) {
  * @param {object} optionsSchema A schema specifying possible options and their types, defaults,
  *   deprecation status, and whether or not they are required.
  * @param {object} providedOptions The options passed into the operation.
- * @param {object} validationOptions Options for the validation itself.
+ * @param {object} overrideOptions Optional. The values for options that need to check inheritance, like readPreference.
```
Comment (mbroadst) [inline lib/options_validator.js]: these are signaled as optional by naming them with brackets around the option name, e.g. 
```
@param {object} [overrideOptions] ...
```

File: lib/options_validator.js [hunk 3]
```diff
@@ -89,6 +95,17 @@ function validationFunction(optionsSchema, providedOptions, validationOptions) {
       const deprecationMessage = `option [${optionName}] is deprecated and will be removed in a later version.`;
       emitDeprecationWarning(deprecationMessage);
     }
+
+    if (overrideOptions && overrideOptions[optionName] != null) {
```
Comment (mbroadst) [inline lib/options_validator.js]: we seem to have lost my last comment, but to reiterate: if we intend these overrides to actually apply in the `validationLevel: none` case, the overrides need to apply first above. You will also need to detect if there is a `default` attempting to overwrite a provided `override` option.

File: lib/options_validator.js [hunk 4]
```diff
@@ -46,48 +52,57 @@ function validationFunction(optionsSchema, providedOptions, validationOptions) {
...
 
-    if (optionFromSchema.required && !verifiedOptions.hasOwnProperty(optionName)) {
-      const invalidateMessageRequired = `required option [${optionName}] was not found.`;
-      invalidate(validationLevel, invalidateMessageRequired, logger);
+    if (overrideOptions && overrideOptions[optionName] != null) {
+      if (!verifiedOptions.hasOwnProperty(optionName)) {
+        verifiedOptions[optionName] = overrideOptions[optionName];
+      }
+      if (optionFromSchema && optionFromSchema.default) {
+        console.warn(
```
Comment (daprahamian) [inline lib/options_validator.js]: I could see this warning confusing users. Something like:

```js
const coll = client.db('foo', { readPreference: ReadPreference.secondary }).collection('bar');
const result = await coll.findOne({...}); // might console.warn if there is a "default" for readPreference
```

Should we be warning here?

Comment (kvwalker) [reply lib/options_validator.js]: In this case, the "default" for `readPreference` would actually be handled by `overrideOptions` so we would have no `default` value for `readPreference`, but we would have `overrideOptions = { readPreference: db.readPreference }`. There shouldn't be anywhere that we have both a `default` and an `overrideOption`.

**Pull request merged**